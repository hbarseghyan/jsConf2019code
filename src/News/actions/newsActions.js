import { getNews } from "../services/newsService";

export const actionTypes = {
  GET_NEWS: "GET_NEWS",
};

export const updateNews = payload => ({
  type: actionTypes.GET_NEWS,
  payload,
});

export const loadNews = () => async dispatch => {
  const res = await getNews();
  dispatch(updateNews(res));
};
