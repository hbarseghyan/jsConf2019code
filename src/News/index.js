import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

import { Container, Row, Col, Spinner } from "react-bootstrap";
import NewsItem from "./components/NewsItem";
import { loadNews } from "./actions/newsActions";
import getNews from "./selectors/newsSelectors";

const News = () => {
  const dispatch = useDispatch();
  const news = useSelector(getNews);

  useEffect(() => {
    dispatch(loadNews());
  }, []);

  return (
    <div className="bootstrap-wrapper content-container">
      <Container>
        <Row>
          <Col>
            <h1 className="page-header">News</h1>
          </Col>
        </Row>
        <Row>
          {news && news.size ? (
            news.map(item => (
              <Col key={item.get("id")}>
                <Row>
                  <NewsItem name={item.get("name")} text={item.get("text")} />
                </Row>
              </Col>
            ))
          ) : (
            <Spinner animation="border" role="status">
              <span className="sr-only">Loading...</span>
            </Spinner>
          )}
        </Row>
      </Container>
    </div>
  );
};

export default News;
