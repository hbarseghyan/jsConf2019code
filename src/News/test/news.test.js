import React from "react";
import * as reactRedux from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { fromJS } from "immutable";
import { shallow, mount } from "enzyme";
import * as newsAction from "../actions/newsActions";
import data from "../../common/resource";

import News from "..";
import { wrap } from "module";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("News", () => {
  describe("News loaded", () => {
    let wrapper;
    beforeEach(() => {
      jest.unmock("react-redux");

      reactRedux.useDispatch = jest.fn().mockImplementation(call => {
        const dispatch = () => call;
        return dispatch;
      });
      reactRedux.useSelector = jest.fn().mockImplementation(() => fromJS(data));
      // jest.spyOn(reactRedux, "useDispatch").mockImplementation(() => fromJS(data));

      wrapper = shallow(<News />);
      console.log(wrapper.debug());
    });

    it("should call useDispatch", () => {
      expect(reactRedux.useDispatch).toHaveBeenCalled();
    });

    it("should call useSelector", () => {
      expect(reactRedux.useSelector).toHaveBeenCalled();
    });

    it("should render a cards", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("showld render 5 cards", () => {
      expect(wrapper.find("NewsItem")).toHaveLength(5);
    });

    it("showld render at list one cards", () => {
      expect(wrapper.exists("NewsItem")).toBe(true);
    });

    it("showld render list with props", () => {
      expect(
        wrapper
          .find("NewsItem")
          .first()
          .prop("text")
      ).toEqual("news text ");
    });
  });
  describe("News loaded with store", () => {
    let wrapper;
    let store;
    beforeEach(() => {
      store = mockStore(
        fromJS({
          news: [],
        })
      );
      store.dispatch = jest.fn();
      jest.unmock("../actions/newsActions");
      newsAction.loadNews = jest.fn();

      wrapper = mount(
        <reactRedux.Provider store={store}>
          <News />
        </reactRedux.Provider>
      );
    });

    it("should call dispatch", () => {
      console.log(wrapper);
      expect(newsAction.loadNews).toHaveBeenCalledTimes(1);
    });
  });

  describe("News preloading", () => {
    let wrapper;
    beforeEach(() => {
      jest.unmock("react-redux");
      reactRedux.useDispatch = jest.fn();
      reactRedux.useSelector = jest.fn();
      wrapper = shallow(<News />);
    });
    it("showld renders preloader", () => {
      expect(wrapper.find(".sr-only").text()).toEqual("Loading...");
    });
  });
});
