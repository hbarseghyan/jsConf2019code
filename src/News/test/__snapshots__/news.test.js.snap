// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News News loaded should render a cards 1`] = `
<div
  className="bootstrap-wrapper content-container"
>
  <Container
    fluid={false}
  >
    <ForwardRef
      noGutters={false}
    >
      <Col>
        <h1
          className="page-header"
        >
          News
        </h1>
      </Col>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <Col
        key="1000"
      >
        <ForwardRef
          noGutters={false}
        >
          <NewsItem
            name="news 1"
            text="news text "
          />
        </ForwardRef>
      </Col>
      <Col
        key="2000"
      >
        <ForwardRef
          noGutters={false}
        >
          <NewsItem
            name="news 2"
            text="news text "
          />
        </ForwardRef>
      </Col>
      <Col
        key="3000"
      >
        <ForwardRef
          noGutters={false}
        >
          <NewsItem
            name="news 3"
            text="news text "
          />
        </ForwardRef>
      </Col>
      <Col
        key="4000"
      >
        <ForwardRef
          noGutters={false}
        >
          <NewsItem
            name="news 4"
            text="news text "
          />
        </ForwardRef>
      </Col>
      <Col
        key="5000"
      >
        <ForwardRef
          noGutters={false}
        >
          <NewsItem
            name="news 5"
            text="news text "
          />
        </ForwardRef>
      </Col>
    </ForwardRef>
  </Container>
</div>
`;
